File: frontend/lib/core/services/super_admin_service.dart
import 'dart:developer';

import '../../models/super_admin_models.dart';
import 'api_service.dart';

/// Service for Super Admin dashboard APIs
/// Handles all super admin related API calls with proper error handling
class SuperAdminService {
  factory SuperAdminService() => _instance;
  SuperAdminService._internal();
  static final SuperAdminService _instance = SuperAdminService._internal();

  final ApiService _apiService = ApiService();

  /// Get complete dashboard data
  /// Returns aggregated data for super admin dashboard
  Future<SuperAdminDashboardResponse> getDashboardData({
    int institutionPage = 1,
    int institutionLimit = 5,
  }) async {
    try {
      log('📊 Fetching super admin dashboard data...');
      
      // The backend dashboard API doesn't accept query parameters
      // It returns a fixed set of data optimized for dashboard
      final response = await _apiService.dio.get('/super-admin/dashboard');

      log('✅ Dashboard data fetched successfully');
      log('🔍 Raw response stats: ${response.data['stats']}');
      
      final dashboardResponse = SuperAdminDashboardResponse.fromJson(response.data);
      log('🔍 Parsed stats: totalInstitutions=${dashboardResponse.stats.totalInstitutions}');
      
      return dashboardResponse;
    } catch (e) {
      log('❌ Error fetching dashboard data: $e');
      rethrow;
    }
  }

  /// Get system statistics only
  /// Returns system-wide statistics for quick overview
  Future<SystemStats> getSystemStats() async {
    try {
      log('📈 Fetching system statistics...');
      
      final response = await _apiService.dio.get('/super-admin/stats');

      log('✅ System stats fetched successfully');
      return SystemStats.fromJson(response.data);
    } catch (e) {
      log('❌ Error fetching system stats: $e');
      rethrow;
    }
  }

  /// Get institution overview with pagination
  /// Returns paginated list of institutions with user counts
  Future<InstitutionListResponse> getInstitutions({
    int page = 1,
    int limit = 10,
    String? search,
    String? status,
    String? type,
  }) async {
    try {
      log('🏢 Fetching institutions (page: $page, limit: $limit)...');
      
      // Convert page to offset for backend API
      final offset = (page - 1) * limit;
      
      final queryParams = <String, dynamic>{
        'limit': limit,
        'offset': offset,
      };

      if (search != null && search.isNotEmpty) {
        queryParams['search'] = search;
      }
      if (status != null && status.isNotEmpty) {
        queryParams['status'] = status;
      }
      if (type != null && type.isNotEmpty) {
        queryParams['type'] = type;
      }

      final response = await _apiService.dio.get(
        '/super-admin/institutions',
        queryParameters: queryParams,
      );

      log('✅ Institutions fetched successfully');
      return InstitutionListResponse.fromJson(response.data);
    } catch (e) {
      log('❌ Error fetching institutions: $e');
      rethrow;
    }
  }

  /// Get user growth trends
  /// Returns monthly user growth data for the last 12 months
  Future<List<UserGrowthTrend>> getUserGrowthTrends({
    int months = 12,
  }) async {
    try {
      log('📊 Fetching user growth trends...');
      
      final response = await _apiService.dio.get(
        '/super-admin/user-growth',
        queryParameters: {
          'months': months,
        },
      );

      log('✅ User growth trends fetched successfully');
      final List<dynamic> data = response.data;
      return data.map((json) => UserGrowthTrend.fromJson(json)).toList();
    } catch (e) {
      log('❌ Error fetching user growth trends: $e');
      rethrow;
    }
  }

  /// Get recent activity
  /// Returns recent system activity for activity feed
  Future<List<RecentActivity>> getRecentActivity({
    int limit = 20,
    int days = 7,
  }) async {
    try {
      log('📝 Fetching recent activity...');
      
      final response = await _apiService.dio.get(
        '/super-admin/recent-activity',
        queryParameters: {
          'limit': limit,
          'days': days,
        },
      );

      log('✅ Recent activity fetched successfully');
      final List<dynamic> data = response.data;
      return data.map((json) => RecentActivity.fromJson(json)).toList();
    } catch (e) {
      log('❌ Error fetching recent activity: $e');
      rethrow;
    }
  }

  /// Get detailed institution information
  /// Returns detailed stats for a specific institution
  Future<InstitutionOverview> getInstitutionDetails(int institutionId) async {
    try {
      log('🔍 Fetching institution details for ID: $institutionId...');
      
      final response = await _apiService.dio.get(
        '/super-admin/institutions/$institutionId',
      );

      log('✅ Institution details fetched successfully');
      return InstitutionOverview.fromJson(response.data);
    } catch (e) {
      log('❌ Error fetching institution details: $e');
      rethrow;
    }
  }

  /// Get system health metrics
  /// Returns aggregated health metrics for monitoring
  Future<Map<String, dynamic>> getSystemHealth() async {
    try {
      log('🏥 Fetching system health metrics...');
      
      // Use system stats as health metrics since there's no separate health endpoint
      final response = await _apiService.dio.get('/super-admin/stats');

      log('✅ System health metrics fetched successfully');
      return response.data;
    } catch (e) {
      log('❌ Error fetching system health: $e');
      rethrow;
    }
  }

  /// Get storage statistics
  /// Returns system storage usage metrics
  Future<Map<String, dynamic>> getStorageStats() async {
    try {
      log('💾 Fetching storage statistics...');
      
      final response = await _apiService.dio.get('/super-admin/storage-stats');

      log('✅ Storage stats fetched successfully');
      return response.data;
    } catch (e) {
      log('❌ Error fetching storage stats: $e');
      rethrow;
    }
  }

  /// Get active sessions count
  /// Returns number of currently active user sessions
  Future<int> getActiveSessionsCount() async {
    try {
      log('👥 Fetching active sessions count...');
      
      final response = await _apiService.dio.get('/super-admin/active-sessions');

      log('✅ Active sessions count fetched successfully');
      return response.data as int;
    } catch (e) {
      log('❌ Error fetching active sessions count: $e');
      rethrow;
    }
  }

  /// Create a new institution
  /// Returns success response with created institution data
  Future<Map<String, dynamic>> createInstitution(Map<String, dynamic> data) async {
    try {
      log('🏢 Creating new institution: ${data['name']}...');
      
      final response = await _apiService.dio.post(
        '/institutions',
        data: data,
      );

      log('✅ Institution created successfully');
      return response.data;
    } catch (e) {
      log('❌ Error creating institution: $e');
      rethrow;
    }
  }

  /// Refresh all dashboard data
  /// Force refresh all cached data (if any caching is implemented)
  Future<void> refreshDashboard() async {
    try {
      log('🔄 Refreshing dashboard data...');
      
      // For now, this just clears any potential cache
      // In the future, this could trigger cache invalidation
      
      log('✅ Dashboard refresh completed');
    } catch (e) {
      log('❌ Error refreshing dashboard: $e');
      rethrow;
    }
  }
}

File: backend/src/super-admin/super-admin.service.ts
import { Injectable, Logger } from '@nestjs/common'
import { PrismaService } from '../prisma/prisma.service'
import {
  InstitutionOverviewQueryDto,
  RecentActivityQueryDto,
  UserGrowthQueryDto,
} from './dto/super-admin.dto'
import {
  InstitutionListResponse,
  InstitutionOverview,
  RecentActivity,
  SuperAdminDashboardResponse,
  SystemStats,
  UserGrowthTrend,
} from './interfaces/super-admin.interfaces'

@Injectable()
export class SuperAdminService {
  private readonly logger = new Logger(SuperAdminService.name)

  constructor(private readonly prisma: PrismaService) {}

  /**
   * Get comprehensive dashboard data for super admin
   * Performance target: < 200ms total
   */
  async getDashboardData(): Promise<SuperAdminDashboardResponse> {
    const startTime = Date.now()

    try {
      // Execute all queries in parallel for optimal performance
      const [stats, institutions, userGrowth, recentActivity] =
        await Promise.all([
          this.getSystemStats(),
          this.getInstitutionOverview({ limit: 5 }), // Top 5 institutions for dashboard
          this.getUserGrowthTrends({ months: 6 }), // Last 6 months for dashboard
          this.getRecentActivity({ limit: 10, days: 7 }), // Last 10 activities
        ])

      const executionTime = Date.now() - startTime
      this.logger.log(`Dashboard data retrieved in ${executionTime}ms`)

      return {
        stats,
        institutions,
        userGrowth,
        recentActivity,
      }
    } catch (error) {
      this.logger.error('Failed to get dashboard data', error)
      throw error
    }
  }

  /**
   * Get system-wide statistics using optimized database view
   * Performance target: < 50ms
   */
  async getSystemStats(): Promise<SystemStats> {
    const startTime = Date.now()

    try {
      const result = await this.prisma.$queryRaw<
        Array<Record<string, unknown>>
      >`
        SELECT * FROM super_admin_system_stats LIMIT 1
      `

      const executionTime = Date.now() - startTime
      this.logger.debug(`System stats retrieved in ${executionTime}ms`)

      if (!result[0]) {
        return this.getEmptySystemStats()
      }

      // Convert BigInt values to numbers for JSON serialization
      const rawStats = result[0]
      const stats: SystemStats = {
        totalInstitutions: Number(rawStats.totalInstitutions || 0),
        inactiveInstitutions: Number(rawStats.inactiveInstitutions || 0),
        totalStudents: Number(rawStats.totalStudents || 0),
        totalTeachers: Number(rawStats.totalTeachers || 0),
        totalAdmins: Number(rawStats.totalAdmins || 0),
        totalStaff: Number(rawStats.totalStaff || 0),
        totalParents: Number(rawStats.totalParents || 0),
        totalActiveUsers: Number(rawStats.totalActiveUsers || 0),
        pendingUsers: Number(rawStats.pendingUsers || 0),
        suspendedUsers: Number(rawStats.suspendedUsers || 0),
        lockedUsers: Number(rawStats.lockedUsers || 0),
        newUsers30d: Number(rawStats.newUsers30d || 0),
        newInstitutions30d: Number(rawStats.newInstitutions30d || 0),
        userHealthPercentage: Number(rawStats.userHealthPercentage || 0),
      }

      return stats
    } catch (error) {
      this.logger.error('Failed to get system stats', error)
      throw error
    }
  }

  /**
   * Get institution overview with pagination and filtering
   * Performance target: < 100ms
   */
  async getInstitutionOverview(
    query: InstitutionOverviewQueryDto = {}
  ): Promise<InstitutionListResponse> {
    const { status, type, limit = 20, offset = 0, search } = query
    const startTime = Date.now()

    try {
      // Build WHERE conditions
      const conditions: string[] = []
      const params: unknown[] = []
      let paramIndex = 1

      if (status) {
        conditions.push(`status = $${paramIndex}`)
        params.push(status)
        paramIndex++
      }

      if (type) {
        conditions.push(`type = $${paramIndex}`)
        params.push(type)
        paramIndex++
      }

      if (search) {
        conditions.push(
          `(name ILIKE $${paramIndex} OR code ILIKE $${paramIndex})`
        )
        params.push(`%${search}%`)
        paramIndex++
      }

      const whereClause =
        conditions.length > 0 ? `WHERE ${conditions.join(' AND ')}` : ''

      // Get total count and data in parallel
      const [totalResult, rawData] = await Promise.all([
        this.prisma.$queryRawUnsafe<Array<{ count: bigint }>>(
          `SELECT COUNT(*) as count FROM super_admin_institution_overview ${whereClause}`,
          ...params
        ),
        this.prisma.$queryRawUnsafe<Array<Record<string, unknown>>>(
          `SELECT * FROM super_admin_institution_overview ${whereClause} 
           ORDER BY "createdAt" DESC 
           LIMIT $${paramIndex} OFFSET $${paramIndex + 1}`,
          ...params,
          limit,
          offset
        ),
      ])

      // Convert BigInt values to numbers for JSON serialization
      const data: InstitutionOverview[] = rawData.map(item => ({
        id: Number(item.id),
        code: String(item.code),
        name: String(item.name),
        type: item.type as 'SCHOOL' | 'COLLEGE' | 'UNIVERSITY' | 'INSTITUTE',
        status: item.status as 'ACTIVE' | 'INACTIVE',
        createdAt: new Date(item.createdAt as string),
        adminName: item.adminName ? String(item.adminName) : null,
        adminEmail: item.adminEmail ? String(item.adminEmail) : null,
        adminStatus: item.adminStatus ? String(item.adminStatus) : null,
        totalUsers: Number(item.totalUsers || 0),
        activeUsers: Number(item.activeUsers || 0),
        students: Number(item.students || 0),
        teachers: Number(item.teachers || 0),
        staff: Number(item.staff || 0),
        parents: Number(item.parents || 0),
        healthPercentage: Number(item.healthPercentage || 0),
      }))

      const total = Number(totalResult[0]?.count || 0)
      const totalPages = Math.ceil(total / limit)
      const page = Math.floor(offset / limit) + 1

      const executionTime = Date.now() - startTime
      this.logger.debug(`Institution overview retrieved in ${executionTime}ms`)

      return {
        data,
        meta: {
          total,
          page,
          limit,
          totalPages,
        },
      }
    } catch (error) {
      this.logger.error('Failed to get institution overview', error)
      throw error
    }
  }

  /**
   * Get user growth trends using optimized database view
   * Performance target: < 75ms
   */
  async getUserGrowthTrends(
    query: UserGrowthQueryDto = {}
  ): Promise<UserGrowthTrend[]> {
    const { months = 12 } = query
    const startTime = Date.now()

    try {
      const rawResult = await this.prisma.$queryRaw<
        Array<Record<string, unknown>>
      >`
        SELECT 
          month,
          "newUsers",
          "activeNewUsers",
          "cumulativeUsers"
        FROM super_admin_user_growth 
        WHERE month >= NOW() - INTERVAL '${months} months'
        ORDER BY month DESC
        LIMIT ${months}
      `

      // Convert BigInt values to numbers for JSON serialization
      const result: UserGrowthTrend[] = rawResult.map(item => ({
        month: new Date(item.month as string),
        newUsers: Number(item.newUsers || 0),
        activeNewUsers: Number(item.activeNewUsers || 0),
        cumulativeUsers: Number(item.cumulativeUsers || 0),
      }))

      const executionTime = Date.now() - startTime
      this.logger.debug(`User growth trends retrieved in ${executionTime}ms`)

      return result
    } catch (error) {
      this.logger.error('Failed to get user growth trends', error)
      throw error
    }
  }

  /**
   * Get recent activity using optimized database view
   * Performance target: < 60ms
   */
  async getRecentActivity(
    query: RecentActivityQueryDto = {}
  ): Promise<RecentActivity[]> {
    const { limit = 20, days = 7 } = query
    const startTime = Date.now()

    try {
      const result = await this.prisma.$queryRaw<RecentActivity[]>`
        SELECT 
          "activityType",
          description,
          timestamp,
          "institutionId"
        FROM super_admin_recent_activity 
        WHERE timestamp >= NOW() - INTERVAL '${days} days'
        ORDER BY timestamp DESC
        LIMIT ${limit}
      `

      const executionTime = Date.now() - startTime
      this.logger.debug(`Recent activity retrieved in ${executionTime}ms`)

      return result
    } catch (error) {
      this.logger.error('Failed to get recent activity', error)
      throw error
    }
  }

  /**
   * Get detailed institution statistics by ID
   * Performance target: < 80ms
   */
  async getInstitutionDetails(
    institutionId: number
  ): Promise<InstitutionOverview | null> {
    const startTime = Date.now()

    try {
      const rawResult = await this.prisma.$queryRaw<
        Array<Record<string, unknown>>
      >`
        SELECT * FROM super_admin_institution_overview 
        WHERE id = ${institutionId}
        LIMIT 1
      `

      const executionTime = Date.now() - startTime
      this.logger.debug(`Institution details retrieved in ${executionTime}ms`)

      if (!rawResult[0]) {
        return null
      }

      // Convert BigInt values to numbers for JSON serialization
      const item = rawResult[0]
      const result: InstitutionOverview = {
        id: Number(item.id),
        code: String(item.code),
        name: String(item.name),
        type: item.type as 'SCHOOL' | 'COLLEGE' | 'UNIVERSITY' | 'INSTITUTE',
        status: item.status as 'ACTIVE' | 'INACTIVE',
        createdAt: new Date(item.createdAt as string),
        adminName: item.adminName ? String(item.adminName) : null,
        adminEmail: item.adminEmail ? String(item.adminEmail) : null,
        adminStatus: item.adminStatus ? String(item.adminStatus) : null,
        totalUsers: Number(item.totalUsers || 0),
        activeUsers: Number(item.activeUsers || 0),
        students: Number(item.students || 0),
        teachers: Number(item.teachers || 0),
        staff: Number(item.staff || 0),
        parents: Number(item.parents || 0),
        healthPercentage: Number(item.healthPercentage || 0),
      }

      return result
    } catch (error) {
      this.logger.error('Failed to get institution details', error)
      throw error
    }
  }

  /**
   * Get storage usage statistics (placeholder for future implementation)
   * Performance target: < 30ms
   */
  async getStorageStats(): Promise<{
    used: string
    total: string
    percentage: number
  }> {
    // TODO: Implement actual storage calculation
    // For now, return mock data that matches the current frontend
    return {
      used: '0 GB',
      total: '100 GB',
      percentage: 0,
    }
  }

  /**
   * Get active sessions count (placeholder for future implementation)
   * Performance target: < 20ms
   */
  async getActiveSessionsCount(): Promise<number> {
    // TODO: Implement actual session tracking
    // For now, return count of users logged in today
    try {
      const result = await this.prisma.$queryRaw<Array<{ count: bigint }>>`
        SELECT COUNT(*) as count 
        FROM users 
        WHERE last_login >= CURRENT_DATE 
        AND status = 'ACTIVE'
      `
      return Number(result[0]?.count || 0)
    } catch (error) {
      this.logger.error('Failed to get active sessions count', error)
      return 0
    }
  }

  /**
   * Helper method to return empty system stats
   */
  private getEmptySystemStats(): SystemStats {
    return {
      totalInstitutions: 0,
      inactiveInstitutions: 0,
      totalStudents: 0,
      totalTeachers: 0,
      totalAdmins: 0,
      totalStaff: 0,
      totalParents: 0,
      totalActiveUsers: 0,
      pendingUsers: 0,
      suspendedUsers: 0,
      lockedUsers: 0,
      newUsers30d: 0,
      newInstitutions30d: 0,
      userHealthPercentage: 0,
    }
  }
}
